doc/code/curves.md

```diff
@@ -0,0 +1,217 @@
+As a consequence, curves are not copy constructable as they require a unique ID for
+event management. However, it is possible to copy the keyframes of one curve to
+the other using the `sync(..)` method. `sync(..)` also works for curves with different
+values types if a converter function from one value type to the other is supplied.
+
+#### Common Operations
+
+All primitive curves support the following operations. They may work slightly different
+for specific curve types.
+
+**Read**
+
+| Method          | Description                                                 |
+| --------------- | ----------------------------------------------------------- |
+| `get(t)`        | Get (interpolated) value at time `t`                        |
+| `frame(t)`      | Get the previous keyframe (time and value) before or at `t` |
+| `next_frame(t)` | Get the next keyframe (time and value) after `t`            |
+
+**Modify**
+
+| Method                  | Description                                                                       |
+| ----------------------- | --------------------------------------------------------------------------------- |
+| `set_insert(t, value)`  | Insert a new keyframe value at time `t`                                           |
+| `set_last(t, value)`    | Insert a new keyframe value at time `t`; delete all keyframes after time `t`      |
+| `set_replace(t, value)` | Insert a new keyframe value at time `t`; remove all other keyframes with time `t` |
+
+**Copy**
+
+| Method           | Description                                                                            |
+| ---------------- | -------------------------------------------------------------------------------------- |
+| `sync(Curve, t)` | copy keyframes from `Curve`, starting at time `t`; delete all keyframes after time `t` |
+
+
+#### Discrete
+
+![Discrete curve function example](images/discrete_curve.png)
+
+Discrete curves implement **constant interpolation** between keyframes. This means
+that the value returned by `get(t)` is always equal to the value of the previous
+keyframe.
+
+Discrete curves should be used for values that only change at specific points in time,
+e.g. for the health of a game entity.
+
+
+#### Continuous
+
+![Continuous curve function example](images/continuous_curve.png)
+
+Continuous curves implement **linear interpolation** between keyframes. This means
+that the value returned by `get(t)` is calculated from the value difference
+between the keyframes before and after `t`. If there is no keyframe after `t`,
+the value of the previous keyframe is used (like on discrete curves).
+
+Value types on continuous curves need to implement methods for the `operator*(..)` and
+`operator-(..)` operations to support linear interpolation. In particular, `operator*(..)`
+must support multiplication with `time::time_t` and `operator-(..)` must support
+substraction for values of the same type.
```

Comment:
now that we can use concepts, we should check it that way :)

Reply:
a good idea for another PR :D